src/main/sensors/opticalflow.c

```diff
@@ -0,0 +1,200 @@
+
+PG_RESET_TEMPLATE(opticalflowConfig_t, opticalflowConfig,
+    .opticalflow_hardware = OPTICALFLOW_NONE,
+    .rotation = 0,
+    .flow_lpf = 0,
+    .flipY = 0
+);
+
+static opticalflow_t opticalflow;
+float cosRotAngle = 1.0f;
+float sinRotAngle = 0.0f;
+static pt2Filter_t xFlowLpf, yFlowLpf;
+
+// ======================================================================
+// =================== Opticalflow Main Functions =======================
+// ======================================================================
+static bool opticalflowDetect(opticalflowDev_t * dev, uint8_t opticalflowHardwareToUse) {
+    UNUSED(dev);
+
+    opticalflowType_e opticalflowHardware = OPTICALFLOW_NONE;
+    requestedSensors[SENSOR_INDEX_OPTICALFLOW] = opticalflowHardwareToUse;
+
+    switch (opticalflowHardwareToUse) {
+        case OPTICALFLOW_MT:
+#ifdef USE_OPTICALFLOW_MT
+            if (mtOpticalflowDetect(dev, rangefinderConfig()->rangefinder_hardware)) {
+                opticalflowHardware = OPTICALFLOW_MT;
+                rescheduleTask(TASK_OPTICALFLOW, TASK_PERIOD_MS(dev->delayMs));
+            }
+#endif
+            break;
+
+        case OPTICALFLOW_NONE:
+            opticalflowHardware = OPTICALFLOW_NONE;
+            break;
+    }
+
+    if (opticalflowHardware == OPTICALFLOW_NONE) {
+        sensorsClear(SENSOR_OPTICALFLOW);
+        return false;
+    }
+
+    detectedSensors[SENSOR_INDEX_OPTICALFLOW] = opticalflowHardware;
+    sensorsSet(SENSOR_OPTICALFLOW);
+    return true;    
+}
+
+bool opticalflowInit(void) {
+    if (!opticalflowDetect(&opticalflow.dev, opticalflowConfig()->opticalflow_hardware)) {
+        return false;
+    }
+
+    opticalflow.dev.init(&opticalflow.dev);
+    opticalflow.quality = OPTICALFLOW_NO_NEW_DATA;
+    opticalflow.rawFlowRates.X = 0;
+    opticalflow.rawFlowRates.Y = 0;
+    opticalflow.processedFlowRates.X = 0;
+    opticalflow.processedFlowRates.Y = 0;
+    opticalflow.lastValidResponseTimeMs = millis();
+
+    if (opticalflowConfig()->rotation != 0) {
+        cosRotAngle = cosf(DEGREES_TO_RADIANS(opticalflowConfig()->rotation));
+        sinRotAngle = sinf(DEGREES_TO_RADIANS(opticalflowConfig()->rotation));
+    }
+    //low pass filter
+    if (opticalflowConfig()->flow_lpf != 0) {
+        const float flowCutoffHz = (float)opticalflowConfig()->flow_lpf / 100.0f;
+        const float flowGain     = pt2FilterGain(flowCutoffHz, opticalflow.dev.delayMs / 1000.0f);
+
+        pt2FilterInit(&xFlowLpf, flowGain);
+        pt2FilterInit(&yFlowLpf, flowGain);
+    }
+    return true;
+}
+
+void opticalflowUpdate(void) {
+    if (opticalflow.dev.update) {
+        opticalflow.dev.update(&opticalflow.dev);
+    }
+}
+
+void opticalflowProcess(void) {
+    opticalflowData_t data = {0};
+    bool hasNewData = opticalflow.dev.read(&opticalflow.dev, &data);
+    
+    opticalflow.quality = data.quality;
+    data.devMinQualityThreshold = opticalflow.dev.minQualityThreshold;
```

Comment:
Why is this set repeatedly?

Reply:
The quality is related to every reading, so it should be updated, the opticalflow.dev.minQualityThreshold is stupidness from me, and I removed it